### PR TITLE
Add RL-based CQL recommender

### PR DESCRIPTION
**Pull request by pkuderov**
Conservative Q-Learning (CQL) algorithm is a SAC-based data-driven deep reinforcement learning algorithm, which achieves state-of-the-art performance in offline RL problems. For discrete-space problems it is based on Double-DQN.

For this request a [d3rlpy](https://github.com/takuseno/d3rlpy) implementation of a CQL is used. It was adapted to work with PySpark dataframes. Hence, I had to add new dependecy to `pyproject.toml`.

Besides an implementation itself, I supply PR with a notebook (and bare python script) to compare CQL with several baseline recommenders the same way it is done in [./experiments/02_models_comparison.ipynb](https://github.com/sb-ai-lab/RePlay/blob/main/experiments/02_models_comparison.ipynb). It is tested on the same ML1M dataset with an exact same seed. Results are the following:

CQL  comparison results

NB: I have problems to update `poetry.lock` as I develop on Mac M1, which requires me to loosen up the dependencies to setup and run RePlay. So, I don't have an exact dev environment as it is specified in `poetry.lock`. Which is why I cannot easily commit the needed changes.

### CODE REVIEW

File: replay/models/cql.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+import torch
+import torch.nn as nn
+from d3rlpy.dataset import MDPDataset
+from pandas import DataFrame
+
+from replay.data_preparator import DataPreparator
+from replay.models.base_torch_rec import TorchRecommender
+
+
+class CQL(TorchRecommender):
```
Comment (shashist) [inline replay/models/cql.py:21]: consider inheriting from Recommender since you are not using the Torch Recommender functionality

File: replay/models/cql.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+    
+    def __init__(
+        self, *,
+        k: int, n_epochs: int = 1,
+        **params
+    ):
+        super().__init__()
+        self.k = k
+        self.n_epochs = n_epochs
+        self.model = CQL_d3rlpy.CQL(**params)
```
Comment (shashist) [inline replay/models/cql.py:124]: what about specifying parameters explicitly?

File: replay/models/cql.py [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+        reward_scaler (d3rlpy.preprocessing.RewardScaler or str):
+            reward preprocessor. The available options are
+            ``['clip', 'min_max', 'standard']``.
+        impl (d3rlpy.algos.torch.cql_impl.CQLImpl): algorithm implementation.
+
+    """
+
+    k: int
+    n_epochs: int
+
```
Comment (shashist) [inline replay/models/cql.py:115]: there should also be `_search_space` with arguments needed

File: replay/models/cql.py [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+            actions = np.array(df_s['relevance'] +
+                               0.1 * np.random.randn(len(df_s)))[:, None],
+            rewards = df_s['rewards'],
+            terminals = df_s['terminals']
+        )
+        return train_dataset
+    
+    @property
+    def _init_args(self):
+        pass
```
Comment (shashist) [inline replay/models/cql.py:193]: please complete this function

File: replay/models/cql.py [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+        pandas_df: pd.DataFrame,
+        model: nn.Module,
+        items_np: np.ndarray,
+        k: int,
+        item_count: int,
+    ) -> pd.DataFrame:
+        pass
+
+    @staticmethod
+    def _predict_by_user_pairs(
```
Comment (shashist) [inline replay/models/cql.py:212]: we need `predict_pairs` method, so when inheriting from `Recommender` please add `_predict_pairs`

**Comment by shashist** [comment]
could you also add tests for the new model?

**Comment by shashist** [comment]
consider pulling recent updates, we have fixed ci/cd in https://github.com/sb-ai-lab/RePlay/commit/6a2eaa01c2c221e91e444b0533f7290cf406051f

File: replay/models/cql.py [hunk 6]
```diff
@@ -0,0 +1,354 @@
+"""
+Using CQL implementation from `d3rlpy` package.
+For 'alpha' version PySpark DataFrame are converted to Pandas
```
Comment (shashist) [inline replay/models/cql.py:3]: Do we still need this line?

File: replay/models/cql.py [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+            soft_q_backup=soft_q_backup,
+            use_gpu=use_gpu,
+            scaler=scaler,
+            action_scaler=action_scaler,
+            reward_scaler=reward_scaler,
+            **params
+        )
+
+    @staticmethod
+    def _rate_user_items(
```
Comment (shashist) [inline replay/models/cql.py:208]: How about TorchRecommender-like naming? CQL class isn't inherited from it but quite related

File: replay/models/cql.py [hunk 8]
```diff
@@ -0,0 +1,354 @@
...
+        ).float().cpu()
+
+        with io.BytesIO(model) as buffer:
+            model = torch.jit.load(buffer, map_location=torch.device('cpu'))
+
+        with torch.no_grad():
+            user_item_pairs['relevance'] = model.forward(input_batch).numpy()
+
+        # It doesn't explicitly filter seen items and doesn't return top k items.
+        # Instead, it returns all predictions as is to be filtered further by base methods
```
Comment (shashist) [inline replay/models/cql.py:228]: Clean this comments, please

Comment (pkuderov) [reply replay/models/cql.py]: moved this comment to `_predict` and made it clearer

File: replay/models/cql.py [hunk 9]
```diff
@@ -0,0 +1,357 @@
...
+        # CQL inner params
+        actor_learning_rate: float = 1e-4,
+        critic_learning_rate: float = 3e-4,
+        temp_learning_rate: float = 1e-4,
+        alpha_learning_rate: float = 1e-4,
+        actor_optim_factory: OptimizerFactory = AdamFactory(),
+        critic_optim_factory: OptimizerFactory = AdamFactory(),
+        temp_optim_factory: OptimizerFactory = AdamFactory(),
+        alpha_optim_factory: OptimizerFactory = AdamFactory(),
+        actor_encoder_factory: EncoderArg = "default",
```
Comment (AleXXL1) [inline replay/models/cql.py:148]: It should to describe allowed values for str params like  actor_encoder_factory,q_func_factory and so on

File: replay/models/cql.py [hunk 10]
```diff
@@ -0,0 +1,357 @@
...
+        self,
+        log: DataFrame,
+        user_features: Optional[DataFrame] = None,
+        item_features: Optional[DataFrame] = None,
+    ) -> None:
+        train: MDPDataset = self._prepare_data(log)
+        self.model.fit(train, n_epochs=self.n_epochs)
+
+    def _prepare_data(self, log: DataFrame) -> MDPDataset:
+        user_logs = log.toPandas().sort_values(['user_idx', 'timestamp'], ascending=True)
```
Comment (AleXXL1) [inline replay/models/cql.py:216]: What are you sorting for? If the next action is another sorting

File: replay/models/cql.py [hunk 11]
```diff
@@ -0,0 +1,357 @@
...
+        terminals[user_terminal_idxs] = 1
+
+        # cannot set zero scale as d3rlpy will treat transitions then as discrete
+        assert self.action_randomization_scale > 0
+        action_randomization_scale = self.action_randomization_scale
+        action_randomization = np.random.randn(len(user_logs)) * action_randomization_scale
+
+        actions = user_logs['relevance'].to_numpy().astype(float) + action_randomization
+
+        train_dataset = MDPDataset(
```
Comment (AleXXL1) [inline replay/models/cql.py:246]: you can do first part on Spark and only here start to use Pandas

File: replay/models/cql.py [hunk 12]
```diff
@@ -0,0 +1,357 @@
...
+    def serialize_policy(self) -> bytes:
+        # store using temporary file and immediately read serialized version
+        with tempfile.NamedTemporaryFile(suffix='.pt') as tmp:
+            # noinspection PyProtectedMember
+            self.model._impl.save_policy(tmp.name)
+            with open(tmp.name, 'rb') as policy_file:
+                return policy_file.read()
+
+    @staticmethod
+    def assert_omp_single_thread():
```
Comment (AleXXL1) [inline replay/models/cql.py:349]: i think you this method hasn't to be here. it is common method  for all models with pytorch. Please add it to utils

**Review by AleXXL1 — CHANGES_REQUESTED** [review]

File: replay/models/cql.py (hunk 11 above)
Comment (pkuderov) [reply replay/models/cql.py:246]: Well, it feels in the spirit of `RePlay` to use Spark here. However, I still see more reasons to make it non-distributedly, because:

This process is relatively cheap — traverse, sorting, grouping, no costly joins. The resulting data in terms of size is equal to the input data. Using distributed calculations in my opinion will only degrade performance for the vast majority of the input size range while also complicating the code itself.

I understand my points are somewhat speculative, but still.

File: replay/models/cql.py (hunk 10 above)
Comment (pkuderov) [reply replay/models/cql.py:216]: Such pre-sorting gets us the naturally ordered log grouped by users, which is used further on everywhere — to make `observations`, `actions` and `terminals`. Another sorting is used only to get `top K` items for each user.

File: replay/models/cql.py (hunk 11 above)
Comment (AleXXL1) [reply replay/models/cql.py:246]: it is may be true for small datasets, but not for large datasets.

Could you tell me, why the fit time is so large?

File: tests/models/test_cql.py [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+import pytest
+from pyspark.sql import DataFrame
+
+from replay.constants import LOG_SCHEMA
+from replay.models import CQL
+from tests.utils import spark
+
+
+@pytest.fixture
+def log(spark):
```
Comment (monkey0head) [inline tests/models/test_cql.py:13]: Hi! Thanks for collaboration! I will add comments on the model testing.

We already have some fixtures in tests/utils, e.g. log, log2, long_log_with_features to use in tests.  Please reuse them here. 

We also have some tests in tests/models/test_all_models. Consider adding your model to some of those tests, e.g. your may test your model works as expected in case of new users/items appearance on inference (test_predict_cold_and_new_filter_out and test_predict_new_users/test_predict_cold_users if applicable for your model).  

If you added model saving/loading functionality, please test it in tests/models/test_save_load_models.py.

File: replay/models/cql.py (hunk 12 above)
Comment (pkuderov) [reply replay/models/cql.py]: Done!

File: replay/models/cql.py (hunk 9 above)
Comment (pkuderov) [reply replay/models/cql.py]: Added to the class init args description corresponding lists of available options

File: tests/models/test_cql.py (hunk 13 above)
Comment (pkuderov) [reply tests/models/test_cql.py]: Thank you for point out to these test files. Added our model to the list of input parameters for applicable tests.

**Comment by pkuderov** [comment]
Here's the wrap up for the changes introduced in the last update:

- [x] Implement and test save/load mechanics:
  - I added d3rlpy CQL model initialization to the CQL init. As d3rlpy has a pretty inflexible save/load pipeline, I had to extract/re-implement several necessary stages like getting init params and their deserialization.
  - [x] Test saving/loading in `tests/models/test_save_load_models`

- [x] Reuse fixtures in `tests/utils` for custom CQL tests.

- [x] Add CQL to tests in `tests/models/test_all_models`
  - [x] Add to `test_predict_cold_and_new_filter_out`, `test_predict_new_users`, `test_predict_cold_users` if applicable. Yes, they are applicable. However, the current stage of the model implementation is not particularly fit for cold predictions. We didn't even test the performance. But it is expected to be done in future updates.
- [x] Check test coverage. pytest+cov report 93% coverage for CQL.py
- [x] Move assert_omp_single_thread to utils
- [x] Investigate why `fit` time is so large. It's due to the number of epochs used for learning. CQL is still noticeably slower than classic recommenders, however GPU acceleration makes execution time acceptable. We have not optimized performance yet (e.g. the scaling of the network size relative to the dataset size). The MDP dataset preparation takes just fractions of the entire fit time.
- [x] Switch `_prepare_data` to use spark
- [x] Describe `actor_encoder_factory`, `q_func_factory` and the other similar init args.

**Comment by pkuderov** [comment]
Moved an MDP dataset preparation (its implementation) outside CQL method implementation while keeping the building itself inside the `_fit` method. This way the thoroughly tested recommender's `fit` pipeline is kept intact, while at the same time provides control to the user on how an MDP dataset should be built.

File: replay/models/cql.py [hunk 14]
```diff
@@ -131,9 +132,8 @@ class CQL(Recommender):
         impl (d3rlpy.algos.torch.cql_impl.CQLImpl): algorithm implementation.
```
Comment (shashist) [inline replay/models/cql.py:131]: Do we have this argument?

File: replay/models/cql.py [hunk 15]
```diff
@@ -154,10 +153,8 @@ class CQL(Recommender):
     
     def __init__(
```
Comment (shashist) [inline replay/models/cql.py:153]: Could you please move stable variables to class variables? Probably `actor_optim_factory, critic_optim_factory, temp_optim_factory, alpha_optim_factory, actor_encoder_factory, critic_encoder_factory, q_func_factory, n_frames, n_steps, n_critics, initial_temperature, initial_alpha, alpha_threshold, n_action_samples, soft_q_backup, scaler, reward_scaler`

**Comment by shashist** [comment]
Do we need @staticmethod for `_get_model_hyperparams`? Seems like it will only we used with `self.model`

File: experiments/cql_compare.py [hunk 16]
```diff
@@ -15,6 +15,7 @@
 
 from replay.data_preparator import DataPreparator, Indexer
 from replay.experiment import Experiment
+from replay.mdp_dataset_builder import MdpDatasetBuilder
```
Comment (shashist) [inline experiments/cql_compare.py:18]: What about placing `MdpDatasetBuilder` in cql.py (before CQL class for example)? Code in `replay` directory  typically model-independent. In case of positive response keep in mind moving `tests/test_mdp_dataset_builder.py` to `tests/models/test_cql.py`

File: replay/models/cql.py [hunk 17]
```diff
@@ -372,12 +333,11 @@ def grouped_map(user_log: pd.DataFrame) -> pd.DataFrame:
...
         # non-model hyperparams
         args = dict(
-            top_k=self.top_k,
-            action_randomization_scale=self.action_randomization_scale,
             n_epochs=self.n_epochs,
+            mdp_dataset_builder=self.mdp_dataset_builder.init_args(),
         )
         # model internal hyperparams
-        args.update(**self._get_model_hyperparams(self.model))
+        args |= self._get_model_hyperparams(self.model)
```
[the remark this answers is not in the document]
Comment (pkuderov) [reply replay/models/cql.py:339]: Yes, `|=` is a dict update operator — both versions are the same :)

File: replay/models/cql.py (hunk 14 above)
Comment (pkuderov) [reply replay/models/cql.py:131]: It can be accepted by the inner implementation, so yes.

File: replay/models/cql.py [hunk 18]
```diff
@@ -131,9 +132,8 @@ class CQL(Recommender):
         impl (d3rlpy.algos.torch.cql_impl.CQLImpl): algorithm implementation.
     """
 
-    top_k: int
     n_epochs: int
-    action_randomization_scale: float
+    mdp_dataset_builder: MdpDatasetBuilder
```
[the remark this answers is not in the document]
Comment (pkuderov) [reply replay/models/cql.py]: I probably do not follow what you mean. Could you clarify? It's just a type hint declaration — I don't see it's duplicated anywhere.

File: replay/models/cql.py (hunk 15 above)
Comment (pkuderov) [reply replay/models/cql.py:158]: Honestly, I am not sure it's a good idea make them class variables since they aren't expected to be read-only for instances, which makes changing them error-prone.

File: replay/models/cql.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (shashist) [reply replay/models/cql.py]: `|=` is not working in python 3.8. We prefer to keep compatibility, 3.8 is officially supported until 10.2024

[the remark this answers is not in the document]
Comment (pkuderov) [reply replay/models/cql.py]: ah, forgot it! gotcha

**Review by AleXXL1 — APPROVED** [review]
